### ISSUE TITLE
Why doesn't base_model subclass nn.module?

### ISSUE THREAD

**Issue by himat** [junyanz/pytorch-CycleGAN-and-pix2pix#590]
I'm trying to understand the design decision behind why the `base_model` and thus none of the models here actually inherit from `nn.module` like normal pytorch models do that I've seen in most examples.

I feel like that would just be more of a normal thing to do, since you could still add all the other methods that the current `base_model` has like set_input, optimize_parameters, update_learning_rate, etc. Is there actually some drawback or hard implementation challenge that happens if you subclass `nn.module`? 

Would love to know the thought process here, thanks!

**Comment by junyanz** [comment]
I think the model aims to implement the formulation (e.g., pix2pix/cyclegan) rather than a network architecture. That's why we don't use `nn.module`. We implemented the code during the first release of PyTorch 0.1, when `nn.module` was not so popular.  What's your opinion (cons vs. pros)?  @SsnL

**Comment by himat** [comment]
Related question: why did you make a different `initialize` method for each model class instead of just using the default `__init__`?

@SsnL

**Comment by SsnL** [comment]
The `Model` class used here include multiple networks (`Module`s), optimizers for the networks, and learning rate schedulers for the optimizers. So it is really more than what `nn.Module` should do. It is more like an owner of the `Module`s, rather than is a `Module` it self.

In the current codebase, `initialize` could very well just be put in `__init__`. However, one could see the possibility of needing to reinitialize an existing model. In that case it will be helpful.
